zaza/openstack/charm_tests/designate/tests.py

```diff
@@ -0,0 +1,197 @@
+        for dom in self.zones_list():
+            if isinstance(dom, dict):
+                if dom['name'] == domain_name:
+                    domain_id = dom['name']
+                    break
+            else:
+                if dom.name == domain_name:
+                    domain_id = dom.id
+                    break
+        return domain_id
+
+    def _get_test_domain_id(self):
+        return self._get_domain_id(self.TEST_DOMAIN)
+
+    @tenacity.retry(
+        wait=tenacity.wait_exponential(multiplier=1, min=5, max=10),
+        reraise=True
+    )
+    def _wait_on_domain_gone(self):
+        logging.debug('Waiting for domain to disappear')
+        if self._get_test_domain_id():
+            raise Exception("Domain Exists")
+
+    @tenacity.retry(
+        wait=tenacity.wait_exponential(multiplier=1, min=5, max=10),
+        reraise=True
+    )
+    def _wait_to_resolve_test_record(self):
+        dns_ip = zaza_juju.get_relation_from_unit(
+            'designate/0',
+            'designate-bind/0',
+            'dns-backend'
+        ).get('private-address')
+
+        logging.info('Waiting for dns record to propagate @ {}'.format(dns_ip))
+        lookup_cmd = [
+            'dig', '+short', '@{}'.format(dns_ip),
+            self.TEST_WWW_RECORD]
+        cmd_out = subprocess.check_output(
+            lookup_cmd, universal_newlines=True).rstrip()
+        if not self.TEST_RECORD[self.TEST_WWW_RECORD] == cmd_out:
+            raise Exception("Record Doesn't Exist")
+
+    def test_400_domain_creation(self):
+        """Simple api calls to create domain."""
+        logging.debug('Checking if domain exists before trying to create it')
+        old_dom_id = self._get_test_domain_id()
+        if old_dom_id:
+            logging.debug('Deleting old domain')
+            self.zones_delete(old_dom_id)
+            self._wait_on_domain_gone()
+
+        logging.debug('Creating new domain')
+        domain = domains.Domain(
+            name=self.TEST_DOMAIN,
+            email="fred@amuletexample.com")
+
+        if self.post_xenial_queens:
+            new_domain = self.designate.zones.create(
+                name=domain.name, email=domain.email)
+        else:
+            new_domain = self.designate.domains.create(domain)
+        self.assertIsNotNone(new_domain)
+
+        logging.debug('Creating new test record')
+        _record = records.Record(
+            name=self.TEST_WWW_RECORD,
+            type="A",
+            data=self.TEST_RECORD[self.TEST_WWW_RECORD])
+
+        if self.post_xenial_queens:
+            _domain_id = new_domain['id']
+            self.designate.recordsets.create(
+                _domain_id, _record.name, _record.type, [_record.data])
+        else:
+            _domain_id = new_domain.id
+            self.designate.records.create(_domain_id, _record)
+
+        self._wait_to_resolve_test_record()
+
+        logging.debug('Tidy up delete test record')
+        self.zones_delete(_domain_id)
```

Comment:
I'd wait for the zone to be deleted as well; at some point this test could contaminate another test.

Reply:
@ajkavanagh, i'm now confused at this point about this test. 

Here's the [original](https://opendev.org/openstack/charm-designate-bind/src/branch/master/src/tests/basic_deployment.py#L274).  it seems likes we are deletings the zone, but not the domain?  Is that because the domain is contained within the zone?  Does that hold true for cases where zones weren't available (pre-xenial-queens)

Reply:
It was more that the one further up in this test function looks like this:

            self.zones_delete(old_dom_id)
            self._wait_on_domain_gone()

i.e. wait until it is gone.  It would be useful if this test ensured that it cleaned up after itself (which it does), but it exits *before* it is sure that the test record is gone.  I'd just think that the equivalent of `_wait_on_domain_gone()` for the new domain that's been created should be done before the function exits.  My reasoning is that the function ought to be idempotent with respect to the resources it creates/removes.

Reply:
I can't tell.  Does it wait on `_domain_id` which is `new_domain['id']` and is that different or the same as `old_dom_id` above; if different, then does it actually wait?  if the same then why a different variable name?

Reply:
honestly... i don't know.  i'll try to understand what the original amulet test was trying to accomplish as much of this structure remains as a straight port of old tests.

Reply:
Okay, I appreciate this is complicated!

Reply:
@ajkavanagh -- i think i cracked it.  Lemme know what you think